Detectors/FIT/FDD/simulation/src/Digitizer.cxx

```diff
@@ -100,92 +143,172 @@ void Digitizer::process(const std::vector<o2::fdd::Hit>* hits, o2::fdd::Digit* d
     }
-    //MCLabels
-    Int_t parentID = hit.GetTrackID();
-    if (parentID != parent) {
-      o2::fdd::MCLabel label(hit.GetTrackID(), mEventID, mSrcID, pmt);
-      if (mMCLabels)
-        mMCLabels->addElement(mMCLabels->getIndexedSize(), label);
-      parent = parentID;
-    } //labels
-  }   //hit loop
-
-  //Conversion of analogue pulse shape to values provided by FEE
-  for (Int_t ipmt = 0; ipmt < parameters.mNchannels; ++ipmt) {
-    channel_data[ipmt].mTime = SimulateTimeCFD(ipmt);
-    for (Int_t iBin = 0; iBin < mNBins; ++iBin)
-      channel_data[ipmt].mChargeADC += mTime[ipmt][iBin] / parameters.mChargePerADC;
-    // LOG(INFO) << "ADC " << channel_data[ipmt].mChargeADC << " Time " << channel_data[ipmt].mTime;
   }
 }
 //_____________________________________________________________________________
-Float_t Digitizer::SimulateTimeCFD(Int_t channel)
+void Digitizer::flush(std::vector<o2::fdd::Digit>& digitsBC,
+                      std::vector<o2::fdd::ChannelData>& digitsCh,
+                      o2::dataformats::MCTruthContainer<o2::fdd::MCLabel>& labels)
+{
+
+  // do we have something to flush? We can do this only for cached BC data which is distanced from currently processed BC by NBCReadOut
+  int nCached = mCache.size();
+  if (nCached < 1) {
+    return;
+  }
+  if (mIntRecord.differenceInBC(mCache.back()) > -BCCacheMin) {
+    LOG(DEBUG) << "Generating new pedestal BL fluct. for BC range " << mCache.front() << " : " << mCache.back();
+    //generatePedestal();
+  } else {
+    return;
+  }
+  //o2::InteractionRecord ir0(mCache.front());
+  //int cacheSpan = 1 + mCache.back().differenceInBC(ir0);
+  //LOG(INFO) << "Cache spans " << cacheSpan << " with " << nCached << " BCs cached";
+
+  for (int ibc = 0; ibc < nCached; ibc++) { // digitize BCs which might not be affected by future events
+    auto& bc = mCache[ibc];
+    storeBC(bc, digitsBC, digitsCh, labels);
+  }
+  // clean cache for BCs which are not needed anymore
+  //LOG(INFO) << "Cleaning cache";
+  mCache.erase(mCache.begin(), mCache.end());
+}
+//_____________________________________________________________________________
+void Digitizer::storeBC(const BCCache& bc,
+                        std::vector<o2::fdd::Digit>& digitsBC, std::vector<o2::fdd::ChannelData>& digitsCh,
+                        o2::dataformats::MCTruthContainer<o2::fdd::MCLabel>& labels)
+{
+  //LOG(INFO) << "Storing BC " << bc;
+
+  int first = digitsCh.size();
+  for (int ic = 0; ic < mNchannels; ic++) {
+    digitsCh.emplace_back(ic, SimulateTimeCFD(bc.pulse[ic]), IntegrateCharge(bc.pulse[ic]), 0, 0, 0, 0, 0, 0, 0, 0, 0);
+  }
+  //bc.print();
+
+  int nBC = digitsBC.size();
+  digitsBC.emplace_back(first, 16, bc, mTriggers);
+
+  for (const auto& lbl : bc.labels)
+    labels.addElement(nBC, lbl);
+}
+
+//_____________________________________________________________________________
+Float_t Digitizer::IntegrateCharge(ChannelBCDataF pulse)
```

Comment:
again, pass the const reference rather than copy the whole pulse array

Reply:
Done